File: kosaraju/golang/kosaraju.go
package kosaraju

import(
	"sort"
)


func DoKosarajuScc(edges [][]int) map[int][]int {
	if edges == nil || len(edges) == 0 {
		return make(map[int][]int)
	}

	g := make(map[int][]int)
	for _, edge := range edges {
		from := edge[0]
		to := edge[1]
		if _, ok :=g[from]; !ok {
			g[from] = []int{}
		}
		g[from] = append(g[from], to)
	}

	isVistited1 := make(map[int]bool)
	finishStack := []int{}
	for v := range g {
		dfs1(g, v, isVistited1, &finishStack)
	}

	rg := revg(g)
	componments := make(map[int][]int)
	isVistited2 := make(map[int]bool)
	for len(finishStack) > 0 {
		v := finishStack[len(finishStack)-1]
		finishStack = finishStack[0:len(finishStack)-1]

		componment := []int{}
		dfs2(rg, v, isVistited2, &componment)

		if len(componment) > 0 {
			sort.Ints(componment)
			componments[componment[0]] = componment
		}
	}

	return componments
}

func dfs1(g map[int][]int, curVertex int, isVistited map[int]bool, finishStack *[]int) {
	if _, ok := isVistited[curVertex]; !ok {
		isVistited[curVertex] = true
		for _, v := range g[curVertex] {
			dfs1(g, v, isVistited, finishStack)
		}
		*finishStack = append(*finishStack, curVertex)
	}
}

func revg(g map[int][]int) map[int][]int {
	revg := make(map[int][]int)

	for fromV, toVs := range g {
		for _, toV := range toVs {
			if _,ok:=revg[toV];!ok {
				revg[toV] = []int{}
			}

			revg[toV] = append(revg[toV], fromV)
		}
	}

	return revg
}

func dfs2(g map[int][]int, curVertex int, isVistited map[int]bool, componment *[]int) {
	if _,ok:=isVistited[curVertex];!ok {
		isVistited[curVertex] = true
		for _, v := range g[curVertex] {
			dfs2(g, v, isVistited, componment)
		}
		*componment = append(*componment, curVertex)
	}
}

File: kosaraju/python3/kosaraju.py
def do_kosaraju_scc(edges):
    if edges == None or len(edges) == 0:
        return {}

    g = {}
    for edge in edges:
        from_v = edge[0]
        to_v = edge[1]
        if from_v not in g:
            g[from_v] = []
        if to_v not in g:
            g[to_v] = []
        g[from_v].append(to_v)

    is_vistited1 = {}
    finish_stack = []
    for v in g:
        _dfs1(g, v, is_vistited1, finish_stack)

    rev_g = _rever_g(g)

    is_vistited2 = {}
    componments = {}
    while len(finish_stack) > 0:
        v = finish_stack[-1]
        finish_stack = finish_stack[0:-1]
        componment = []
        _dfs2(rev_g, v, is_vistited2, componment)

        if len(componment) > 0:
            componment.sort()
            componments[componment[0]] = componment
    
    return componments
        




def _dfs1(g, v, is_vistited, finish_stack):
    if v not in is_vistited:
        is_vistited[v] = True
        for to_v in g[v]:
            _dfs1(g, to_v, is_vistited, finish_stack)
        finish_stack.append(v)

def _rever_g(g):
    rev_g = {}
    for from_v in g:
        if from_v not in rev_g:
            rev_g[from_v] = []
        for to_v in g[from_v]:
            if to_v not in rev_g:
                rev_g[to_v] = []
            rev_g[to_v].append(from_v) 
    return rev_g
        
def _dfs2(g, v, is_vistited, componment):
    if v not in is_vistited:
        is_vistited[v] = True
        for to_v in g[v]:
            _dfs2(g, to_v, is_vistited, componment)
        componment.append(v)





